pkg/controllers/management/aks/aks_cluster_handler.go

```diff
@@ -0,0 +1,723 @@
+		noNodePoolsOnUpstreamSpec := len(cluster.Status.AKSStatus.UpstreamSpec.NodePools) == 0
+		if (cluster.Spec.AKSConfig.NodePools != nil && noNodePoolsOnSpec) || (cluster.Spec.AKSConfig.NodePools == nil && noNodePoolsOnUpstreamSpec) {
+			cluster, err = e.setFalse(cluster, apimgmtv3.ClusterConditionWaiting, addNgMessage)
+			if err != nil {
+				return cluster, err
+			}
+		} else {
+			if apimgmtv3.ClusterConditionWaiting.GetMessage(cluster) == addNgMessage {
+				cluster = cluster.DeepCopy()
+				apimgmtv3.ClusterConditionWaiting.Message(cluster, "Waiting for API to be available")
+				cluster, err = e.clusterClient.Update(cluster)
+				if err != nil {
+					return cluster, err
+				}
+			}
+		}
+
+		cluster, err = e.setTrue(cluster, apimgmtv3.ClusterConditionProvisioned, "")
+		if err != nil {
+			return cluster, err
+		}
+
+		// If there are no subnets it can be assumed that networking fields are not provided. In which case they
+		// should be created by the aks-operator, and needs to be copied to the cluster object.
+		if len(cluster.Status.AKSStatus.Subnets) == 0 {
+			subnets, _ := status["subnets"].([]interface{})
+			if len(subnets) != 0 {
+				// network field have been generated and are ready to be copied
+				virtualNetwork, _ := status["virtualNetwork"].(string)
+				subnets, _ := status["subnets"].([]interface{})
+				cluster = cluster.DeepCopy()
+
+				// change fields on status to not be generated
+				cluster.Status.AKSStatus.VirtualNetwork = virtualNetwork
+				for _, val := range subnets {
+					cluster.Status.AKSStatus.Subnets = append(cluster.Status.AKSStatus.Subnets, val.(string))
+				}
+				cluster, err = e.clusterClient.Update(cluster)
+				if err != nil {
+					return cluster, err
+				}
+			}
+		}
+		if cluster.Status.APIEndpoint == "" {
+			return e.recordCAAndAPIEndpoint(cluster)
+		}
+
+		if !cluster.Status.AKSStatus.UpstreamSpec.PublicAccess {
+			// Check to see if we can still use the public API endpoint even though
+			// the cluster has private-only access
+			serviceToken, mustTunnel, err := e.generateSATokenWithPublicAPI(cluster)
+			if mustTunnel != nil {
+				cluster = cluster.DeepCopy()
+				cluster.Status.ServiceAccountToken = serviceToken
+				return e.clusterClient.Update(cluster)
+			}
+			if err != nil {
+				return cluster, err
+			}
+		}
+
+		if cluster.Status.ServiceAccountToken == "" {
+			cluster, err = e.generateAndSetServiceAccount(cluster)
+			if err != nil {
+				var statusErr error
+				if strings.Contains(err.Error(), fmt.Sprintf(dialer.WaitForAgentError, cluster.Name)) {
+					// In this case, the API endpoint is private and rancher is waiting for the import cluster command to be run.
+					cluster, statusErr = e.setUnknown(cluster, apimgmtv3.ClusterConditionWaiting, "waiting for cluster agent to be deployed")
+					if statusErr == nil {
+						e.clusterEnqueueAfter(cluster.Name, enqueueTime)
+					}
+					return cluster, statusErr
```

Comment:
Same as above, is there a private setup that this would apply to?

Reply:
Yes, it will be private cluster setup.